### PR TITLE
Tweak `r_arg_as_*()` error messages

### PR DESCRIPTION
**Pull request by DavisVaughan**
As noted in https://github.com/r-lib/vctrs/pull/1616#discussion_r944609452

### CODE REVIEW

**Review by lionel- — APPROVED** [review]
Should we export `r_obj_type_friendly()` via C callable?

File: src/rlang/vec.h [hunk 1]
```diff
@@ -339,7 +339,7 @@ bool r_is_string(r_obj* x) {
 static inline
 bool r_arg_as_bool(r_obj* x, const char* arg) {
   if (!r_is_bool(x)) {
-    r_abort("`%s` must be a logical value.", arg);
+    r_abort("`%s` must be `TRUE` or `FALSE`.", arg);
```
Comment (lionel-) [inline src/rlang/vec.h:342]: By the way the rlang lib exports `r_format_lazy_error_arg()`, this returns a vmax-protected C string that is equivalent to `{.arg }` formatting.

**Comment by DavisVaughan** [comment]
> r_obj_type_friendly

That sounds useful! Currently our C errors are less useful than ones thrown by things like `rlang:::check_bool()` because of that

**Comment by lionel-** [comment]
I've opened https://github.com/r-lib/rlang/issues/1463

**Pull request merged**